docs/configuration.md
Configuration Reference
=======================

Here you will find all configuration options of `sylius_grid`.

<details open><summary>Yaml</summary>

```yaml
sylius_grid:
    grids:
        app_user: # Your grid name
            driver:
                name: doctrine/orm
                options:
                    class: "%app.model.user.class%"
                    repository:
                        method: myCustomMethod
                        arguments:
                            id: resource.id
            sorting:
                name: asc
            limits: [10, 25, 50, 100]
            fields:
                name:
                    type: twig # Type of field
                    label: Name # Label
                    path: . # dot means a whole object
                    sortable: ~ | field path
                    position: 100
                    options:
                        template: "@Grid/Column/_name.html.twig" # Only twig column
                        vars:
                            labels: # a template of how does the label look like
                    enabled: true
            filters:
                name:
                    type: string # Type of filter
                    label: app.ui.name
                    enabled: true
                    template: ~
                    position: 100
                    options:
                        fields: { }
                    form_options:
                        type: contains # type of string filtering option, if you one to have just one
                    default_value: ~
                enabled:
                    type: boolean # Type of filter
                    label: app.ui.enabled
                    enabled: true
                    template: ~
                    position: 100
                    options:
                        field: enabled
                    form_options: { }
                    default_value: ~
                date:
                    type: date # Type of filter
                    label: app.ui.created_at
                    enabled: true
                    template: ~
                    position: 100
                    options:
                        field: createdAt
                    form_options: { }
                    default_value: ~
                channel:
                    type: entity # Type of filter
                    label: app.ui.channel
                    enabled: true
                    template: ~
                    position: 100
                    options:
                        fields: [channel]
                    form_options:
                        class: "%app.model.channel.class%"
                    default_value: ~
            actions:
                main:
                    create:
                        type: create
                        label: sylius.ui.create
                        enabled: true
                        icon: ~
                        position: 100
                item:
                    update:
                        type: update
                        label: sylius.ui.edit
                        enabled: true
                        icon: ~
                        position: 100
                        options: { }
                    delete:
                        type: delete
                        label: sylius.ui.delete
                        enabled: true
                        icon: ~
                        position: 100
                        options: { }
                    show:
                        type: show
                        label: sylius.ui.show
                        enabled: true
                        icon: ~
                        position: 100
                        options:
                            link:
                                route: app_user_show
                                parameters:
                                    id: resource.id
                    archive:
                        type: archive
                        label: sylius.ui.archive
                        enabled: true
                        icon: ~
                        position: 100
                        options:
                            restore_label: sylius.ui.restore
                bulk:
                    delete:
                        type: delete
                        label: sylius.ui.delete
                        enabled: true
                        icon: ~
                        position: 100
                        options: { }
                subitem:
                    addresses:
                        type: links
                        label: sylius.ui.manage_addresses
                        options:
                            icon: cubes
                            links:
                                index:
                                    label: sylius.ui.list_addresses
                                    icon: list
                                    route: app_admin_user_address_index
                                    visible: resource.hasAddress
                                    parameters:
                                        userId: resource.id
                                create:
                                    label: sylius.ui.generate
                                    icon: random
                                    route: app_admin_user_address_create
                                    parameters:
                                        userId: resource.id
```

</details>

<details open><summary>PHP</summary>

```php
<?php

use Sylius\Bundle\GridBundle\Builder\Action\Action;
use Sylius\Bundle\GridBundle\Builder\ActionGroup\BulkActionGroup;
use Sylius\Bundle\GridBundle\Builder\ActionGroup\ItemActionGroup;
use Sylius\Bundle\GridBundle\Builder\ActionGroup\MainActionGroup;
use Sylius\Bundle\GridBundle\Builder\ActionGroup\SubItemActionGroup;
use Sylius\Bundle\GridBundle\Builder\Field\Field;
use Sylius\Bundle\GridBundle\Builder\Filter\Filter;
use Sylius\Bundle\GridBundle\Builder\GridBuilder;
use Sylius\Bundle\GridBundle\Config\GridConfig;

return static function (GridConfig $grid): void {
    $grid->addGrid(GridBuilder::create('app_user', '%app.model.user.class%')
        ->setDriver('doctrine/orm')
        ->setRepositoryMethod('myCustomMethod', ['id' => 'resource.id'])
        ->orderBy('name', 'asc')
        ->setLimits([10, 25, 50, 100])
        ->addField(
            Field::create('name', 'twig') // Name & Type of field
                ->setLabel('Name') // # Label
                ->setPath('.') // dot means a whole object
                ->setSortable(true)
                ->setPosition(100)
                ->setOptions([
                    'template' => '@Grid/Column/_name.html.twig', // Only twig column
                ])
                ->setEnabled(true)
        )
        ->addFilter(
            Filter::create('name', 'string') // Name & Type of filter
                ->setLabel('app.ui.name')
                ->setEnabled(true)
                ->setOptions(['fields' => []])
                ->setFormOptions(['type' => 'contains']) // type of string filtering option, if you one to have just one
        )
        ->addActionGroup(MainActionGroup::create(
            Action::create('create', 'create')
                ->setLabel('sylius.ui.create')
                ->setEnabled(true)
                ->setIcon('plus')
                ->setPosition(100)
                ->setOptions([]),
        ))
        ->addActionGroup(
            ItemActionGroup::create(
                Action::create('update', 'update')
                    ->setLabel('sylius.ui.edit')
                    ->setEnabled(true)
                    ->setIcon('pencil')
                    ->setPosition(100)
                    ->setOptions([]),
                Action::create('delete', 'delete')
                    ->setLabel('sylius.ui.delete')
                    ->setEnabled(true)
                    ->setIcon('trash')
                    ->setPosition(100)
                    ->setOptions([]),
                Action::create('show', 'show')
                    ->setLabel('sylius.ui.show')
                    ->setEnabled(true)
                    ->setIcon('search')
                    ->setPosition(100)
                    ->setOptions([
                        'link' => [
                            'route' => 'app_user_show',
                            'parameters' => [
                                'id' => 'resource.id',
                            ],          
                        ],          
                    ]),
                Action::create('archive', 'archive')
                    ->setLabel('sylius.ui.archive')
                    ->setEnabled(true)
                    ->setIcon('search')
                    ->setPosition(100)
                    ->setOptions([
                        'restore_label' => 'sylius.ui.restore',          
                    ]),
            )
        )
        ->addActionGroup(
            BulkActionGroup::create(
                Action::create('delete', 'delete')
                    ->setLabel('sylius.ui.delete')
                    ->setEnabled(true)
                    ->setIcon('trash')
                    ->setPosition(100)
                    ->setOptions([]),
            )
        )
        ->addActionGroup(
            SubItemActionGroup::create(
                Action::create('addresses', 'links')
                    ->setLabel('sylius.ui.manage_addresses')
                    ->setOptions([
                        'icon' => 'cubes',
                        'links' => [
                            'index' => [
                                'label' => 'sylius.ui.list_addresses',
                                'icon' => 'list',
                                'route' => 'app_admin_user_address_index',
                                'visible' => 'resource.hasAddress',
                                'parameters' => [
                                    'userId' => 'resource.id',
                                ],
                            ],
                            'create' => [
                                'label' => 'sylius.ui.generate',
                                'icon' => 'random',
                                'route' => 'app_admin_user_address_create',
                                'parameters' => [
                                    'userId' => 'resource.id',
                                ],
                            ],
                        ],
                    ]),
            )
        )
    );
};
```

</details>


docs/field_configuration.md
Field configuration
===================

Each field can be configured with several configuration keys, to make it
more suitable to your grid requirements.

  | Name     | Type   | Description                                                                                         |
  |----------|--------|-----------------------------------------------------------------------------------------------------|
  | type     | string | Type of column. Default field types are described [here](field_types.md).                           |
  | label    | string | Label displayed in the field header. By default, it is field name.                                  |
  | path     | string | Path to property displayed in field (can be property of resource or one of its referenced objects). |
  | position | int    | Position of field in the grid index view.                                                           |
  | options  | array  | Array of field options (see below).                                                                 |

`options` field can contains following fields:

  | Name     | Type   | Description                                                                                                 | Default     |
  |----------|--------|-------------------------------------------------------------------------------------------------------------|-------------|
  | template | string | Available (and required) only for *twig* column type. Path to template that is used to render column value. |             |
  | format   | string | Available only for *datetime* field type.                                                                   | `Y:m:d` `H:i:s` |


CHANGELOG.md
# CHANGELOG

## CHANGELOG FOR `1.11.x`

### v1.11.0-RC.1 (2022-03-08)

#### Details

- [#214](https://github.com/Sylius/SyliusGridBundle/issues/214) [Docs] Resorting pages on docs homepage ([@loic425](https://github.com/loic425))
- [#215](https://github.com/Sylius/SyliusGridBundle/issues/215) [Docs] Add docs for the grid maker ([@loic425](https://github.com/loic425))
- [#216](https://github.com/Sylius/SyliusGridBundle/issues/216) [Docs] Add documentation for Grids as service ([@loic425](https://github.com/loic425))
- [#218](https://github.com/Sylius/SyliusGridBundle/issues/218) [Docs] Fix the template reference ([@Roshyo](https://github.com/Roshyo))
- [#219](https://github.com/Sylius/SyliusGridBundle/issues/219) [DX] [Docs] Adding more typing to FieldType and in doc ([@Roshyo](https://github.com/Roshyo))

### v1.11.0-BETA.1 (2022-02-10)

#### Details

- [#209](https://github.com/Sylius/SyliusGridBundle/issues/209) Hotfix with datasources ([@loic425](https://github.com/loic425))
- [#211](https://github.com/Sylius/SyliusGridBundle/issues/211) [Fix] Register grid maker only when maker is enabled ([@loic425](https://github.com/loic425))
- [#212](https://github.com/Sylius/SyliusGridBundle/issues/212) [Fix] Add a logic exception to install the needed adapter ([@loic425](https://github.com/loic425))

### v1.11.0-ALPHA.1 (2022-02-02)

#### TL;DR

- GridBuilder configurable in PHP 🎉
- Bump PHP to 8.0 and Symfony to 5.4
- Make grid command

#### Details

- [#119](https://github.com/Sylius/SyliusGridBundle/issues/119) [GridBuilder] Grids as service ([@loic425](https://github.com/loic425))
- [#152](https://github.com/Sylius/SyliusGridBundle/issues/152) Upgrade to GitHub-native Dependabot ([@dependabot-preview](https://github.com/dependabot-preview)[[@bot](https://github.com/bot)])
- [#155](https://github.com/Sylius/SyliusGridBundle/issues/155) Grid config builders ([@loic425](https://github.com/loic425))
- [#163](https://github.com/Sylius/SyliusGridBundle/issues/163) Update phpstan/phpstan-webmozart-assert requirement from 0.12.12 to 0.12.13 ([@dependabot](https://github.com/dependabot)[[@bot](https://github.com/bot)])
- [#169](https://github.com/Sylius/SyliusGridBundle/issues/169) Update vimeo/psalm requirement from 4.6.4 to 4.9.2 ([@dependabot](https://github.com/dependabot)[[@bot](https://github.com/bot)])
- [#173](https://github.com/Sylius/SyliusGridBundle/issues/173) [GridBuilder] Predefined fields ([@loic425](https://github.com/loic425))
- [#174](https://github.com/Sylius/SyliusGridBundle/issues/174) [GridBuilder] Predefined action groups ([@loic425](https://github.com/loic425))
- [#175](https://github.com/Sylius/SyliusGridBundle/issues/175) [GridBuilder] Predefined filters ([@loic425](https://github.com/loic425))
- [#176](https://github.com/Sylius/SyliusGridBundle/issues/176) [GridBuilder] Test PHP config without yaml config and add some missing features ([@loic425](https://github.com/loic425))
- [#177](https://github.com/Sylius/SyliusGridBundle/issues/177) [GridBuilder] Docs ([@loic425](https://github.com/loic425))
- [#178](https://github.com/Sylius/SyliusGridBundle/issues/178) Update vimeo/psalm requirement from 4.6.4 to 4.11.2 ([@dependabot](https://github.com/dependabot)[[@bot](https://github.com/bot)])
- [#179](https://github.com/Sylius/SyliusGridBundle/issues/179) Update phpstan/phpstan-phpunit requirement from 0.12.18 to 0.12.22 ([@dependabot](https://github.com/dependabot)[[@bot](https://github.com/bot)])
- [#180](https://github.com/Sylius/SyliusGridBundle/issues/180) Update phpstan/phpstan requirement from 0.12.94 to 0.12.99 ([@dependabot](https://github.com/dependabot)[[@bot](https://github.com/bot)])
- [#182](https://github.com/Sylius/SyliusGridBundle/issues/182) [GridBuilder] Fix update action label ([@loic425](https://github.com/loic425))
- [#183](https://github.com/Sylius/SyliusGridBundle/issues/183) Tests with Postgres ([@loic425](https://github.com/loic425))
- [#184](https://github.com/Sylius/SyliusGridBundle/issues/184) Fix like with PostgresSQL ([@loic425](https://github.com/loic425))
- [#185](https://github.com/Sylius/SyliusGridBundle/issues/185) [GridBuilder] Fix entity filter with multiple option ([@loic425](https://github.com/loic425))
- [#186](https://github.com/Sylius/SyliusGridBundle/issues/186) [GridBuilder] Add some missing features on field builder ([@loic425](https://github.com/loic425))
- [#187](https://github.com/Sylius/SyliusGridBundle/issues/187) [GridBuilder] Improve documentation for Grid builders ([@loic425](https://github.com/loic425))
- [#188](https://github.com/Sylius/SyliusGridBundle/issues/188) Fixed template syntax in route config ([@veloxy](https://github.com/veloxy))
- [#189](https://github.com/Sylius/SyliusGridBundle/issues/189) Update vimeo/psalm requirement from 4.11.2 to 4.12.0 ([@dependabot](https://github.com/dependabot)[[@bot](https://github.com/bot)])
- [#190](https://github.com/Sylius/SyliusGridBundle/issues/190) [GridBuilder] Allow using class as parameter for grids as service ([@loic425](https://github.com/loic425))
- [#191](https://github.com/Sylius/SyliusGridBundle/issues/191) [GridBuilder] Fix extending a grids as service ([@loic425](https://github.com/loic425))
- [#197](https://github.com/Sylius/SyliusGridBundle/issues/197) Drop support for PHP7.4, Symfony version upgraded to 5.4 ([@Rafikooo](https://github.com/Rafikooo))
- [#200](https://github.com/Sylius/SyliusGridBundle/issues/200) Remove travis build status from README ([@GSadee](https://github.com/GSadee))
- [#202](https://github.com/Sylius/SyliusGridBundle/issues/202) Fix grid builder method positions ([@loic425](https://github.com/loic425))
- [#203](https://github.com/Sylius/SyliusGridBundle/issues/203) Update vimeo/psalm requirement from 4.12.0 to 4.18.1 ([@dependabot](https://github.com/dependabot)[[@bot](https://github.com/bot)])
- [#205](https://github.com/Sylius/SyliusGridBundle/issues/205) Add support for pagerfanta 3 ([@loic425](https://github.com/loic425))

src/Bundle/DependencyInjection/Configuration.php
<?php

/*
 * This file is part of the Sylius package.
 *
 * (c) Paweł Jędrzejewski
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

declare(strict_types=1);

namespace Sylius\Bundle\GridBundle\DependencyInjection;

use Sylius\Bundle\GridBundle\Doctrine\ORM\Driver as DoctrineORMDriver;
use Sylius\Bundle\GridBundle\SyliusGridBundle;
use Symfony\Component\Config\Definition\Builder\ArrayNodeDefinition;
use Symfony\Component\Config\Definition\Builder\TreeBuilder;
use Symfony\Component\Config\Definition\ConfigurationInterface;

final class Configuration implements ConfigurationInterface
{
    public function getConfigTreeBuilder(): TreeBuilder
    {
        $treeBuilder = new TreeBuilder('sylius_grid');

        /** @var ArrayNodeDefinition $rootNode */
        $rootNode = $treeBuilder->getRootNode();

        $this->addDriversSection($rootNode);
        $this->addTemplatesSection($rootNode);
        $this->addGridsSection($rootNode);

        return $treeBuilder;
    }

    private function addDriversSection(ArrayNodeDefinition $node): void
    {
        $node
            ->children()
                ->arrayNode('drivers')
                    ->defaultValue([SyliusGridBundle::DRIVER_DOCTRINE_ORM])
                    ->enumPrototype()->values(SyliusGridBundle::getAvailableDrivers())->end()
                ->end()
            ->end()
        ;
    }

    private function addTemplatesSection(ArrayNodeDefinition $node): void
    {
        $node
            ->children()
                ->arrayNode('templates')
                    ->addDefaultsIfNotSet()
                    ->children()
                        ->arrayNode('filter')
                            ->useAttributeAsKey('name')
                            ->scalarPrototype()->end()
                        ->end()
                        ->arrayNode('action')
                            ->useAttributeAsKey('name')
                            ->scalarPrototype()->end()
                        ->end()
                        ->arrayNode('bulk_action')
                            ->useAttributeAsKey('name')
                            ->scalarPrototype()->end()
                        ->end()
                    ->end()
                ->end()
            ->end()
        ;
    }

    private function addGridsSection(ArrayNodeDefinition $node): void
    {
        $node
            ->children()
                ->arrayNode('grids')
                    ->useAttributeAsKey('code')
                    ->arrayPrototype()
                        ->children()
                            ->scalarNode('extends')->cannotBeEmpty()->end()
                            ->arrayNode('driver')
                                ->addDefaultsIfNotSet()
                                ->children()
                                    ->scalarNode('name')->cannotBeEmpty()->defaultValue(DoctrineORMDriver::NAME)->end()
                                    ->arrayNode('options')
                                        ->performNoDeepMerging()
                                        ->variablePrototype()->end()
                                        ->defaultValue([])
                                    ->end()
                                ->end()
                            ->end()
                            ->arrayNode('sorting')
                                ->performNoDeepMerging()
                                ->useAttributeAsKey('name')
                                ->enumPrototype()->values(['asc', 'desc'])->cannotBeEmpty()->end()
                            ->end()
                            ->arrayNode('limits')
                                ->performNoDeepMerging()
                                ->integerPrototype()->end()
                                ->defaultValue([10, 25, 50])
                            ->end()
                            ->arrayNode('fields')
                                ->useAttributeAsKey('name')
                                ->arrayPrototype()
                                    ->children()
                                        ->scalarNode('type')->isRequired()->cannotBeEmpty()->end()
                                        ->scalarNode('label')->cannotBeEmpty()->end()
                                        ->scalarNode('path')->cannotBeEmpty()->end()
                                        ->scalarNode('sortable')->end()
                                        ->scalarNode('enabled')->defaultTrue()->end()
                                        ->scalarNode('position')->defaultValue(100)->end()
                                        ->arrayNode('options')
                                            ->performNoDeepMerging()
                                            ->variablePrototype()->end()
                                        ->end()
                                    ->end()
                                ->end()
                            ->end()
                            ->arrayNode('filters')
                                ->useAttributeAsKey('name')
                                ->arrayPrototype()
                                    ->children()
                                        ->scalarNode('type')->isRequired()->cannotBeEmpty()->end()
                                        ->scalarNode('label')->cannotBeEmpty()->end()
                                        ->scalarNode('enabled')->defaultTrue()->end()
                                        ->scalarNode('template')->end()
                                        ->scalarNode('position')->defaultValue(100)->end()
                                        ->arrayNode('options')
                                            ->performNoDeepMerging()
                                            ->variablePrototype()->end()
                                        ->end()
                                        ->arrayNode('form_options')
                                            ->performNoDeepMerging()
                                            ->variablePrototype()->end()
                                        ->end()
                                        ->variableNode('default_value')->end()
                                    ->end()
                                ->end()
                            ->end()
                            ->arrayNode('actions')
                                ->useAttributeAsKey('name')
                                ->arrayPrototype()
                                    ->useAttributeAsKey('name')
                                    ->arrayPrototype()
                                        ->children()
                                            ->scalarNode('type')->isRequired()->end()
                                            ->scalarNode('label')->end()
                                            ->scalarNode('enabled')->defaultTrue()->end()
                                            ->scalarNode('icon')->end()
                                            ->scalarNode('position')->defaultValue(100)->end()
                                            ->arrayNode('options')
                                                ->performNoDeepMerging()
                                                ->variablePrototype()->end()
                                            ->end()
                                        ->end()
                                    ->end()
                                ->end()
                            ->end()
                        ->end()
                    ->end()
                ->end()
            ->end()
        ;
    }
}
